core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala

```diff
@@ -93,6 +95,10 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
     } {
       out.close()
     }
+    indexFile.deleteOnExit()
+    if (!tmp.renameTo(indexFile)) {
+      throw new IOException(s"fail to rename index file $tmp to $indexFile")
```

Comment:
On Thu, Nov 12, 2015 at 8:30 AM, Matei Zaharia <notifications@github.com>
wrote:

> In
> core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala
> <https://github.com/apache/spark/pull/9610#discussion_r44678796>:
>
> > @@ -93,6 +95,10 @@ private[spark] class IndexShuffleBlockResolver(conf: SparkConf) extends ShuffleB
> >      } {
> >        out.close()
> >      }
> > +    indexFile.deleteOnExit()
> > +    if (!tmp.renameTo(indexFile)) {
> > +      throw new IOException(s"fail to rename index file $tmp to $indexFile")
>
> Can you test for this? I think the worry was about different TaskSets
> attempting the same map stage. Imagine that attempt 1 of the stage
> successfully completes a task, and sends back a map output status, but that
> status gets ignored because that stage attempt got cancelled. Attempt 2
> might then fail to send a new status for it.
>
> There seem to be two ways to fix it if this problem can actually occur --
> either add MapOutputStatuses even from failed task sets or mark this new
> task as successful if a file exists.
>
After this PR, the second attempt of same task will return SUCCESS, with
new MapOutputStatus, which could be different than the previous attempt
(having different sizes of partitions), since we does not use the exact
number of size (could be lossy compressed), I think it's fine.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/9610/files#r44678796>.
>



-- 
 - Davies